modules/carto/src/api/basemap.ts

```diff
@@ -0,0 +1,156 @@
+import {
+  applyLayerGroupFilters,
+  fetchBasemapStyle,
+  getStyleUrl,
+  someLayerGroupsDisabled
+} from '../basemap';
+import {APIErrorContext, KeplerMapConfig} from './types';
+
+const CUSTOM_STYLE_ID_PREFIX = 'custom:';
+const DEFAULT_CARTO_STYLE = 'positron';
+const CARTO_MAP_STYLES = ['positron', 'dark-matter', 'voyager'];
+
+export type BasemapType = 'maplibre' | 'google-maps';
+
+export type BasemapProps = MaplibreBasemapProps | GoogleBasemapProps;
+
+export type MaplibreBasemapProps = {
+  /**
+   * Type of basemap.
+   */
+  type: 'maplibre';
+
+  /**
+   * Basemap style URL or style object.
+   *
+   * Note,  layers in this style may be filtered by `visibleLayerGroups`.
+   *
+   * Non-filtered pristine version is stored in `rawStyle` property.
+   */
+  style: string | Record<string, any>;
+
+  /**
+   * Layer groups to be displayed in the basemap.
+   */
+  visibleLayerGroups?: Record<string, boolean>;
+
+  /** If `style` has been filtered by `visibleLayerGroups` then this property contains original style object, so user
+   * can use `applyLayerGroupFilters` again with new settings.
+   */
+  rawStyle?: string | Record<string, any>;
+
+  /**
+   * Custom attribution for style data if not provided by style definition.
+   */
+  attribution?: string;
+};
+
+export type GoogleBasemapProps = {
+  /**
+   * Type of basemap.
+   */
+  type: 'google-maps';
+
+  /**
+   * Google map options.
+   */
+  options: Record<string, any>;
```

Comment:
Is the `options` parent object important here?  I think I'd expected it to be flattened out like the maplibre options ...

```javascript
export type GoogleBasemapProps = {
  type: 'google-maps';
  mapTypeId: string;
  mapId?: string;
}
```

... but if there's a reason not to do so, that's fine!

Reply:
I wanted to save us from need to cherry-pick "our" vs" basmap provider props.
Here:
```
type: 'google-maps'; // this is ours and shouldn't go to google maps api
  mapTypeId: string; // those two are actually opaque to this api ... and should go directly to google maps api
  mapId?: string;
``

So in this proposition we _do_ need to cherry-pick props and know they names in advance and take care not make name collisions.

That's why my idea this "basemap" structure only embeds google options or "maplibre" style definitions.

Reply:
Ok, that makes sense – thanks!